homeassistant/components/mqtt/discovery.py

```diff
@@ -102,6 +118,151 @@ def async_log_discovery_origin_info(
+
+    if CONF_AVAILABILITY in discovery_payload:
+        for availability_conf in cv.ensure_list(discovery_payload[CONF_AVAILABILITY]):
+            _replace_abbreviations(availability_conf, ABBREVIATIONS, ABBREVIATIONS_SET)
+
+
+@callback
+def _replace_topic_base(discovery_payload: dict[str, Any]) -> None:
+    """Replace topic base in MQTT discovery data."""
+    base = discovery_payload.pop(TOPIC_BASE)
+    for key, value in discovery_payload.items():
+        if isinstance(value, str) and value:
+            if value[0] == TOPIC_BASE and key.endswith("topic"):
+                discovery_payload[key] = f"{base}{value[1:]}"
+            if value[-1] == TOPIC_BASE and key.endswith("topic"):
+                discovery_payload[key] = f"{value[:-1]}{base}"
+    if discovery_payload.get(CONF_AVAILABILITY):
+        for availability_conf in cv.ensure_list(discovery_payload[CONF_AVAILABILITY]):
+            if not isinstance(availability_conf, dict):
+                continue
+            if topic := str(availability_conf.get(CONF_TOPIC)):
+                if topic[0] == TOPIC_BASE:
+                    availability_conf[CONF_TOPIC] = f"{base}{topic[1:]}"
+                if topic[-1] == TOPIC_BASE:
+                    availability_conf[CONF_TOPIC] = f"{topic[:-1]}{base}"
+
+
+@callback
+def _generate_device_cleanup_config(
+    hass: HomeAssistant, object_id: str, node_id: str | None
+) -> dict[str, Any]:
+    """Generate a cleanup message on device cleanup."""
+    mqtt_data = hass.data[DATA_MQTT]
+    device_node_id: str = f"{node_id} {object_id}" if node_id else object_id
+    config: dict[str, Any] = {CONF_DEVICE: {}, CONF_COMPONENTS: {}}
+    comp_config = config[CONF_COMPONENTS]
+    for platform, discover_id in mqtt_data.discovery_already_discovered:
+        ids = discover_id.split(" ")
+        component_node_id = ids.pop(0)
+        component_object_id = " ".join(ids)
+        if not ids:
+            continue
+        if device_node_id == component_node_id:
+            comp_config[component_object_id] = {CONF_PLATFORM: platform}
+
+    return config if comp_config else {}
+
+
+@callback
+def _parse_device_payload(
+    hass: HomeAssistant,
+    payload: ReceivePayloadType,
+    object_id: str,
+    node_id: str | None,
+) -> dict[str, Any]:
+    """Parse a device discovery payload."""
+    device_payload: dict[str, Any] = {}
+    if payload == "":
+        if not (
+            device_payload := _generate_device_cleanup_config(hass, object_id, node_id)
+        ):
+            _LOGGER.warning(
+                "No device components to cleanup for %s, node_id '%s'",
+                object_id,
+                node_id,
+            )
+        return device_payload
+    try:
+        device_payload = MQTTDiscoveryPayload(json_loads_object(payload))
+    except ValueError:
+        _LOGGER.warning("Unable to parse JSON %s: '%s'", object_id, payload)
+        return {}
+    try:
+        _replace_all_abbreviations(device_payload)
```

Comment:
AFAICT, this won't ever raise `vol.Invalid` so it can be moved ahead of the try block